### PR TITLE
E2E Testing

### PR DESCRIPTION
**Pull request by JuliusKarl**

### CODE REVIEW

File: research-hub-web/cypress/integration/equipment.e2e.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    // });
+
+    // No Organisations in equipment
+    // 
+    // it('displays a list of organisations', () => {
+    //     cy.get('#organisations').contains('Centre for eResearch').should('exist');
+    // });
+
+    // Cypress doesn't dupport multi-tab testing
+    //
```
Comment (rosemcc) [inline research-hub-web/cypress/integration/equipment.e2e.js:60]: some options here: https://filiphric.com/opening-a-new-tab-in-cypress 
really we just want to check link is valid? So you could use something like this as suggested in that article: 
    cy
      .request(link)
      .its('status')
      .should('eq', 200);
  });

Comment (rosemcc) [reply research-hub-web/cypress/integration/equipment.e2e.js:60]: Having said that.. Cam was planning to have a link checker service that will be checking all our external links are valid, so maybe doing that in our e2e tests is unnecessary

Comment (rosemcc) [reply research-hub-web/cypress/integration/equipment.e2e.js:60]: Maybe the best thing is just to check that there is actually a link present

File: research-hub-web/cypress/integration/equipment.e2e.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    it('displays a list of contacts', () => {
+        cy.get('#contacts').contains('Vicki Ormond').should('exist');
+    });
+    
+    // No document in equipment
+    //
+    // it('displays a list of documents', () => {
+    //     cy.get('#documents').contains('IT Acceptable Use Policy').should('exist');
+    // });
```
Comment (rosemcc) [inline research-hub-web/cypress/integration/equipment.e2e.js:46]: related docs and related orgs is in the content model. But we can address these once we create some test content. Then we can make sure the test content contains all possible fields

**Comment by rosemcc** [comment]
Nice Julius! e2e all passing. It's pretty slow and CPU/memory intensive - is that what's causing the issues with Jenkins? There was an error that came up after running the software pages tests, maybe related to memory usage:
`(node:10608) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 preprocessor:close listeners added to [EventEmitter]. Use
emitter.setMaxListeners() to increase limit`

**Comment by rosemcc** [comment]
The only thing I'm not super keen on with the e2e tests is the ones where we are testing for specific text to be present - as soon as anyone makes a change in the content wording those could break. Again its something that would be solved by having a proper set of test entries..

**Comment by rosemcc** [comment]
Unit tests:
Error: AppComponent should create FAILED
        TypeError: Cannot read property 'unsubscribe' of undefined
--> need to make sure you have try/catch around all unsubscribes in ngOnDestroy() (a few components need this - app, search-bar, search-page)

--> make sure you destroy the components after each test:
afterEach(() => {
    fixture.destroy();
});

--> It would be good to clean up some of the 'X is not a known element' errors that are coming up. They're usually caused by not importing or declaring component dependencies. It's good practice to mock dependencies where possible.

**Comment by JuliusKarl** [comment]
> Nice Julius! e2e all passing. It's pretty slow and CPU/memory intensive - is that what's causing the issues with Jenkins? There was an error that came up after running the software pages tests, maybe related to memory usage:
> `(node:10608) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 preprocessor:close listeners added to [EventEmitter]. Use emitter.setMaxListeners() to increase limit`

Yeah that's what i'm thinking is causing it since it's similar to the issue I was running into before

**Comment by JuliusKarl** [comment]
> The only thing I'm not super keen on with the e2e tests is the ones where we are testing for specific text to be present - as soon as anyone makes a change in the content wording those could break. Again its something that would be solved by having a proper set of test entries..

either test entry or test-environment in contentful should help resolve testing changes

**Comment by JuliusKarl** [comment]
> Unit tests:
> Error: AppComponent should create FAILED
> TypeError: Cannot read property 'unsubscribe' of undefined
> --> need to make sure you have try/catch around all unsubscribes in ngOnDestroy() (a few components need this - app, search-bar, search-page)
> 
> --> make sure you destroy the components after each test:
> afterEach(() => {
> fixture.destroy();
> });
> 
> --> It would be good to clean up some of the 'X is not a known element' errors that are coming up. They're usually caused by not importing or declaring component dependencies. It's good practice to mock dependencies where possible.

Yup my mistake for leaving that it, unit tests aren't finished yet

**Comment by rosemcc** [comment]
e2e memory things seems an ongoing open issue: https://github.com/cypress-io/cypress/issues/1305

**Comment by rosemcc** [comment]
Some e2e tests failing - Article and Events pages tests. Everything else is all good 👍

**Comment by rosemcc** [comment]
> Some e2e tests failing - Article and Events pages tests. Everything else is all good 👍

The article and event those tests use are not loading due to the type error issue - so not sure if it's worth changing the tests seeing it should be fixed once we deal with the type errors.

**Pull request merged**